### PR TITLE
Add typescript

### PR DESCRIPTION
**Pull request by jbuget**
### Problem

JavaScript provides some sugar features around OOP with a basic system of classes. But it is very limited, in particular when used in a Domain Layered Architecture which relies a lot on IoC/DI with port/adapters patterns.

### Solution

[TypeScript](https://www.typescriptlang.org/) is an extension of the JavaScript language with an advanced set of OOP features such as Static Type Checking, Interfaces & Generics, Scopes (private, protected, public). It also comes with a handy set of cool tools and community contributions.

This PR is an effort to migrate the API codebase from JS to TS.

### Implementation

**Before**

`api/lib/domain/AccountRepository.js`

```javascript
class AccountRepository {

  save(account) {
    console.log('Returns an Account');
    throw new Error('You must implement this method');
  }

  findById(id) {
    console.log('Returns an Account');
    throw new Error('You must implement this method');
  }

  findByUuid(uuid) {
    console.log('Returns an Account');
    throw new Error('You must implement this method');
  }

  findAccountWithEncryptedPasswordByEmail(email) {
    console.log('Returns an AccountWithEncryptedPassword');
    throw new Error('You must implement this method');
  }

  delete(id) {
    console.log('Returns nothing');
    throw new Error('You must implement this method');
  }

  deleteByUuid(uuid) {
    console.log('Returns nothing');
    throw new Error('You must implement this method');
  }

  existsById(id) {
    console.log('Returns a boolean');
    throw new Error('You must implement this method');
  }

  existsByUuid(uuid) {
    console.log('Returns a boolean');
    throw new Error('You must implement this method');
  }

  isUsernameAvailable(username) {
    console.log('Returns a boolean');
    throw new Error('You must implement this method');
  }

  isEmailAvailable(email) {
    console.log('Returns a boolean');
    throw new Error('You must implement this method');
  }
}
module.exports = AccountRepository;
```

`api/lib/infrastructure/repositories/AccountRepositorySql.js`

```javascript
const AccountRepository = require('../../domain/AccountRepository');
const Account = require('../../domain/Account');
const AccountWithEncryptedPassword = require('../../domain/AccountWithEncryptedPassword');
const { build } = require('./sql-repository-factory');

class AccountRepositorySql extends AccountRepository {

  async findAccountWithEncryptedPasswordByEmail(email) {
    const accountModel = await this.Model.findOne({ where: { email } });
    if (accountModel) {
      return new AccountWithEncryptedPassword(accountModel);
    }
  }

  async isUsernameAvailable(username) {
    const account = await this.Model.findOne({ where: { username } });
    return !account;
  }

  async isEmailAvailable(email) {
    const account = await this.Model.findOne({ where: { email } });
    return !account;
  }

  // ...

}

module.exports = build({
  modelName: 'Account',
  tableName: 'accounts',
  Entity: Account,
  Repository: AccountRepositorySql
});
```

**After**

`api/lib/domain/AccountRepository.ts`

```typescript
import {EntityRepository} from './EntityRepository';
import {Account} from './Account';
import {AccountWithEncryptedPassword} from './AccountWithEncryptedPassword';

interface AccountRepository extends EntityRepository<Account> {
  findAccountWithEncryptedPasswordByEmail(email: string): Promise<AccountWithEncryptedPassword | null>;
  isUsernameAvailable(username: string): Promise<boolean>;
  isEmailAvailable(email: string): Promise<boolean>;
}
export {AccountRepository};
```

`api/lib/infrastructure/repositories/AccountRepositorySql.ts`

```typescript
import {GenericRepositorySql} from './GenericRepositorySql';
import {Account} from '../../domain/Account';
import {Model} from 'sequelize';
import {AccountWithEncryptedPassword} from '../../domain/AccountWithEncryptedPassword';
import {AccountRepository} from '../../domain/AccountRepository';

const models = require('../../../db/models');

class AccountRepositorySql extends GenericRepositorySql<Account> implements AccountRepository {

  constructor() {
    super(models['Account'], 'Account', 'accounts');
  }

  fromModelToDto(model: Model): Account {
    const attributes = {
      id: model.get('id') as number,
      uuid: model.get('uuid') as string,
      createdAt: model.get('createdAt') as Date,
      updatedAt: model.get('updatedAt') as Date,
      name: model.get('name') as string,
      username: model.get('username') as string,
      email: model.get('email') as string,
    };
    return new Account(attributes);
  }

  async findById(id: number) {
    const model = await this.Model.findByPk(id);
    if (model) {
      return this.fromModelToDto(model);
    }
    return null;
  }

  async findAccountWithEncryptedPasswordByEmail(email: string) {
    const accountModel = await this.Model.findOne({where: {email}});
    if (accountModel) {
      const account: AccountWithEncryptedPassword = this.fromModelToDto(accountModel);
      account.encryptedPassword = accountModel.get('encryptedPassword') as string;
      return account;
    }
    return null;
  }

  async isUsernameAvailable(username: string) {
    const account = await this.Model.findOne({where: {username}});
    return !account;
  }

  async isEmailAvailable(email: string) {
    const account = await this.Model.findOne({where: {email}});
    return !account;
  }

  // ...

}
export {AccountRepositorySql};
```

### CODE REVIEW

File: api/lib/domain/Account.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    this.email = attributes.email;
+    this.createdAt = attributes.createdAt;
+    this.updatedAt = attributes.updatedAt;
+  }
+
+}
+
+export {Account};
+
+
```
Comment (VincentHardouin) [inline api/lib/domain/Account.ts:24]: Il te manque un linter 🤔

File: api/lib/infrastructure/ioc.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    const service = this.serviceMap[serviceName];
+    if (!service) {
+      throw new Error(`Service ${serviceName} not found in IOC container`);
+    }
+    return service;
+  }
+
+}
+
+export { IocContainer};
```
Comment (VincentHardouin) [inline api/lib/infrastructure/ioc.ts]: Same

File: api/lib/infrastructure/repositories/VaultSummaryRepositorySql.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  async listAllUserVaultSummaries(accountId: number) {
+    const data = await models.sequelize.query(`
+SELECT v.*, (
+  SELECT COUNT(*)
+  FROM items
+  WHERE "vaultId"=v.id
+) AS "itemsCount"
+FROM vaults as v
+WHERE "accountId"=${accountId}
```
Comment (VincentHardouin) [inline api/lib/infrastructure/repositories/VaultSummaryRepositorySql.ts:33]: Globalement, pour toutes tes requêtes natives, il faut passer par une preparedStatement pour passer les attributs sur Sequelize cela se fait comme ça : https://stackoverflow.com/a/49243091

File: api/lib/domain/Account.ts (hunk 1 above)
Comment (jbuget) [reply api/lib/domain/Account.ts:24]: Oui :-)

**Pull request merged**